File: LinkedList.cpp
// This file contains basic operations on Linked Lists

#include<iostream>
using namespace std;

class Node {

public:
  int data;
  Node* next;

  Node(int val){
    this->data = val;
    this->next = NULL;
  }
};

class LinkedList{

private:
  Node* head;
  int size;
  
public:
  
  LinkedList(){
    head = NULL;
    size = 0;
  }

  int get_size(){
    return size;
  }
  
  void insertBegin(int val){

    Node *temp = new Node(val);
    temp->data = val;
    temp->next = NULL;

    if(head == NULL)
      head = temp;
    else{
      temp->next = head;
      head = temp;
    }
    size++;
  }
  

  void insertEnd(int val){

    Node *temp = new Node(val);
    Node *t;
    temp->data = val;
    temp->next = NULL;

    if(head == NULL)
      head = temp;
    else{

      t = head;

      while(t->next != NULL)
	t = t->next;

      t->next = temp;
    }

    size++;
  }

  void display(){

    Node *temp = head;

    cout<<endl;
    while(temp != NULL){
      cout<<temp->data<<"->";
      temp = temp->next;
    }

    cout<<"NULL"<<endl<<"Current size:"<<size<<endl;
    
  }

  //This function will insert a value at given position.
  //If the position exceeds number of elements, then it will insert at end of list
  
  void insertAtPos(int val, int pos){

    if(pos <= 1){
      insertBegin(val);
      return;
    }

    if(pos > size){
      insertEnd(val);
      return;
    }
    
    Node *temp = new Node(val);
    Node *t = head;

    if(head == NULL)
      head = temp;
    else{
      
      while(t->next != NULL && --pos > 1)
	t = t->next;

      temp->next = t->next;
      t->next = temp;

    }

    size++;
  }


  int deleteBegin(){

    Node *t = head;
    int val;

    if(head == NULL){
      cout<<"No list exists!";
      return -1;
    }

    val = t->data;

    head = head->next;
    free(t);

    size--;
    return val;
  }


  int deleteEnd(){

    Node *t = head, *p;
    int val;
    
    if(head == NULL){
      cout<<"No list exists!";
      return -1;
    }

    while(t->next->next != NULL)
      t = t->next;
    
    p = t->next;
    t->next = NULL;

    val = p->data;
    
    free(p);

    size--;
    return val;
    
  }

  int deleteByPos(int pos){

    Node *t = head, *p;
    
    if(head == NULL){
      cout<<"No list exists!";
      return -1;
    }

    if(pos > size || pos < 1){
      cout<<"Invalid position!";
      return -1;
    }

    if(pos == 1)
      return deleteBegin();

    if(pos == size)
      return deleteEnd();

    while(t->next->next != NULL && --pos > 1)
	t = t->next;

    p = t->next;
    t->next = t->next->next;
    int val = p->data;
    free(p);
    size--;

    return val;
  }


  int searchByVal(int val){
    Node *t = head;

    int pos = 0;
    
    while(t != NULL){
      if(t->data == val)
	return pos+1;

      t = t->next;
      pos++;
    }

    return -1;
  }

  
  int searchByPos(int pos){

    Node *t = head, *p;
    
    if(head == NULL){
      cout<<"No list exists!";
      return -1;
    }

    if(pos > size || pos < 1){
      cout<<"Invalid position!";
      return -1;
    }

    if(pos == 1)
      return t->data;

    while(t->next->next != NULL && pos-- > 1)
	t = t->next;

    int val = t->data;
 
    return val;

  }

  
  int deleteByVal(int val){

    int pos = searchByVal(val);

    if(pos == -1){
      cout<<"\nElement does not exist!";
      return -1;
    }

    return deleteByPos(pos);
  }

};

void displaymenu(){

  cout<<"\nLinked List Operations:";
  cout<<"\n1. Insert at Beginning\n2. Insert at End\n3. Insert by Position";
  cout<<"\n4. Delete at Beginning\n5. Delete at End\n6. Delete by value\n7. Delete by position";
  cout<<"\n8. Search by value\n9. Search by position\n10. Display list\n11. Display Menu\n12. Exit "<<endl;
}

int main(){

  LinkedList L;

  int ch,val,pos;

  displaymenu();
  
  bool flag = true;
  
  while(flag){
    cout<<"\nEnter your choice: ";
    cin>>ch;

    if(ch == 13)
      break;
    
    switch(ch){
      
    case 1:
      cout<<"\nEnter value: ";
      cin>>val;
      L.insertBegin(val);
      break;
      
    case 2:
      cout<<"\nEnter value: ";
      cin>>val;
      L.insertEnd(val);
      break;
      
    case 3:
      cout<<"\nEnter value: ";
      cin>>val;
      cout<<"\nEnter position: ";
      cin>>pos;
      L.insertAtPos(val, pos);
      break;
      
    case 4:
      val = L.deleteBegin();
      if(val == -1)
	cout<<"\nCannot carry operation!";
      else
	cout<<"\n"<<val<<" value deleted";
      break;
      
    case 5:
      val = L.deleteEnd();
      if(val == -1)
	cout<<"\nCannot carry operation!";
      else
	cout<<"\n"<<val<<" value deleted";
      break;
      
    case 6:
      cout<<"\nEnter Value:";
      cin>>val;
      val = L.deleteByVal(val);
      if(val == -1)
	cout<<"\nCannot carry operation!";
      else
	cout<<"\n"<<val<<" value deleted";
      break;

    case 7:
      cout<<"\nEnter position:";
      cin>>pos;
      val = L.deleteByPos(pos);
      if(val == -1)
	cout<<"\nCannot carry operation!";
      else
	cout<<"\n"<<val<<" value deleted";
      break;


    case 8:
      cout<<"\nEnter value:";
      cin>>val;
      pos = L.searchByVal(val);
      if(pos == -1)
	cout<<"\nValue not found!";
      else
	cout<<"\nValue found at position "<<pos<<endl;
      break;

    case 9:
      cout<<"\nEnter position:";
      cin>>pos;
      val = L.searchByPos(pos);
      if(val == -1)
	cout<<"\nCannot carry operation!";
      else
	cout<<"Value at position is: "<<val;
      break;

    case 10:
      L.display();
      break;

    case 11:
      displaymenu();
      break;

    case 12:
      cout<<"\nExiting!"<<endl;
      flag = false;
      break;
      
    default:
      cout<<"\nEnter valid choice!";
    } //end of switch

  } //end of while
  
}

File: LinkedList.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None


class LinkedList:

    def __init__(self):
        self.size = 0
        self.head = None


    def getSize(self):
        return self.size

    
    def insertBegin(self,val):

        temp = Node(val)

        if self.head == None:
            self.head = temp
        else:
            temp.next = self.head
            self.head = temp

        self.size += 1

        return

    
    def insertEnd(self, val):

        temp = Node(val)
        
        if self.head == None: 
            self.head = temp
            return
        
        t = self.head
        
        while t.next != None:
            t = t.next
        
        t.next = temp
        
        self.size += 1

        return

    
    def insertAtPos(self,val,pos):

        if pos > self.size:
            self.insertEnd(val)
            return

        if pos <= 1:
            self.insertBegin(val)
            return

        temp = Node(val)
        
        if self.head == None:  # No element
            self.head = temp
            self.tail = temp
            self.size += 1
            return

        t = self.head

        while t != None and (pos-1) > 1:
            t = t.next
            pos -= 1

        temp.next = t.next
        t.next = temp
            
        self.size += 1

        return
    
    
    def deleteBegin(self):
        
        if self.head == None:
            print("No list!")
            return
  
        t = self.head
        self.head = self.head.next
        t = None
        
        self.size -= 1
        
        return
        
    
    def deleteEnd(self):
        
        if self.head == None:
            print("No list")
            return
          
        temp = self.head
        
        while temp.next != None:
                temp = temp.next
        
        val = temp.data
        temp = None
        
        self.size -= 1
        
        return val
    
    
    def deleteAtPos(self,pos):
        
        if pos >= self.size:
            self.deleteEnd()
            return

        if pos <= 1:
            self.deleteBegin()
            return
        
        temp = self.head
        
        while temp.next != None and (pos-1) > 1:
            temp = temp.next
            pos -= 1
            
        p = temp.next
        temp.next = temp.next.next
        
        val = p.data
        p.next = None
        p = None
        
        self.size -= 1
        
        return val
    
    
    def searchByVal(self,val):
        
        if self.head == None:
            print("No list!")
            return None
        
        t = self.head
        pos = 1
        
        while t != None:
            if t.data == val:
                return pos
            pos += 1
            t = t.next
        
        return None
        
    def searchByPos(self, pos):
        
        if pos > self.size or pos < 1:
            print("Invalid position")
            return None
        
        t = self.head
        
        while t.next != None and (pos-1) >= 1:
            pos -= 1
            t = t.next
        
        return t.data
        
        
    def deleteByPos(self,pos):
        
        if pos > self.size or pos < 1:
            print("Invalid position")
            return None
        
        if pos == 1:
            val = self.head.data
            self.head = self.head.next
            return val
            
        t = self.head
        
        while t.next != None and (pos-1) > 1:
            t = t.next
            pos -= 1
        
        p = t.next
        t.next = t.next.next
        
        val = p.data
        p.next = None
        p = None
        
        self.size -= 1
        
        return val
    

    def deleteByVal(self,val):
        
        pos = self.searchByVal(val)
        
        if pos != None:
            return self.deleteByPos(pos)
        
        return None


    def display(self):

        if self.head == None:
            print("No list!")
            return

        t = self.head
        
        while t != None:
            print("{}-> ".format(t.data), end = "")
            t = t.next

        print("NULL")
        return
